### PR TITLE
Feature/starting basicauth

### PR DESCRIPTION
**Pull request by slashformotion**
Add a session management system
Add handlers to handle authentication with username/password.

### CODE REVIEW

File: commands/initDatabaseCommands.go [hunk 1]
```diff
@@ -26,4 +26,12 @@ func initDatabaseCommands(cfg *verdeter.VerdeterCommand) {
 	databaseSslmodeKey := "database.sslmode"
 	cfg.GKey(databaseSslmodeKey, verdeter.IsStr, "", "The sslmode to use when connecting to the database server")
 	cfg.SetRequired(databaseSslmodeKey)
+
+	databaseRetryKey := "database.retry"
```
Comment (Zorin95670) [inline commands/initDatabaseCommands.go:30]: database.init.retry

File: commands/initDatabaseCommands.go [hunk 2]
```diff
@@ -26,4 +26,12 @@ func initDatabaseCommands(cfg *verdeter.VerdeterCommand) {
...
+	cfg.GKey(databaseRetryKey, verdeter.IsUint, "", "The number of times badaas tries to establish a connection with the database")
+	cfg.SetDefault(databaseRetryKey, uint(10))
+
+	databaseRetryDurationKey := "database.retryDuration"
```
Comment (Zorin95670) [inline commands/initDatabaseCommands.go:34]: database.init.retryTime

File: services/sessionservice/session.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	"github.com/ditrit/badaas/persistence/models"
+	"github.com/ditrit/badaas/persistence/repository"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+)
+
+// Duration Values
+var (
+	// The duration of a session
+	sessionDuration = 1 * time.Minute
```
Comment (Zorin95670) [inline services/sessionservice/session.go:19]: put this in conf

File: services/sessionservice/session.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+	// Pull interval
+	pullTime = time.Second * 30
```
Comment (Zorin95670) [inline services/sessionservice/session.go:22]: put this in conf

File: services/sessionservice/session.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+		cache:             make(map[uuid.UUID]*models.Session),
+		logger:            logger,
+		sessionRepository: sessionRepository,
+	}
+	sessionService.init()
+	return sessionService
+}
+
+// Return true is expired
+func isExpired(session *models.Session) bool {
```
Comment (Zorin95670) [inline services/sessionservice/session.go:67]: to put in models.Session

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: commands/initDatabaseCommands.go (hunk 2 above)
Comment (slashformotion) [reply commands/initDatabaseCommands.go]: done in the other PR

File: commands/initDatabaseCommands.go (hunk 1 above)
Comment (slashformotion) [reply commands/initDatabaseCommands.go]: done in the other PR

File: persistence/repository/CRUDRepository.go [hunk 6]
```diff
@@ -15,6 +15,7 @@ type CRUDRepository[T models.Tabler, ID any] interface {
 	GetByID(ID) (*T, httperrors.HTTPError)
 	GetAll(...pagination.SortOption) ([]*T, httperrors.HTTPError)
 	Count(squirrel.Sqlizer) (uint, httperrors.HTTPError)
-	Find(squirrel.Sqlizer, pagination.Paginator, ...pagination.SortOption) (*pagination.Page[T], httperrors.HTTPError)
-	Transaction(fn func(CRUDRepository[T, ID]) (any, error)) (any, error)
+	Find(squirrel.Sqlizer, ...pagination.SortOption) ([]*T, httperrors.HTTPError)
+	FindPaginated(squirrel.Sqlizer, pagination.Paginator, ...pagination.SortOption) (*pagination.Page[T], httperrors.HTTPError)
```
Comment (Zorin95670) [inline persistence/repository/CRUDRepository.go:19]: remove and do only one find

File: persistence/repository/CRUDRepositoryImpl.go [hunk 7]
```diff
@@ -36,7 +36,7 @@ func NewCRUDRepository[T models.Tabler, ID any](database *gorm.DB, logger *zap.L
 
 // Run the function passed as parameter, if it returns the error and rollback the transaction.
 // If no error is returned, it commits the transaction and return the interface{} value.
-func (repository *CRUDRepositoryImpl[T, ID]) Transaction(transactionFunction func(CRUDRepository[T, ID]) (any, error)) (any, error) {
+func (repository *CRUDRepositoryImpl[T, ID]) Transaction(transactionFunction func(repo CRUDRepository[T, ID]) (any, error)) (any, httperrors.HTTPError) {
```
Comment (Zorin95670) [inline persistence/repository/CRUDRepositoryImpl.go:39]: repo ?

**Review by Zorin95670 — CHANGES_REQUESTED** [review]
rework on your commit

File: persistence/repository/CRUDRepository.go [hunk 8]
```diff
@@ -13,8 +13,7 @@ type CRUDRepository[T models.Tabler, ID any] interface {
 	Delete(*T) httperrors.HTTPError
 	Save(*T) httperrors.HTTPError
 	GetByID(ID) (*T, httperrors.HTTPError)
-	GetAll(...pagination.SortOption) ([]*T, httperrors.HTTPError)
-	Count(squirrel.Sqlizer) (uint, httperrors.HTTPError)
```
Comment (Zorin95670) [inline persistence/repository/CRUDRepository.go:17]: Why do you delete it ?

File: services/sessionservice/session.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+			zap.Uint("userID", session.UserID),
+			zap.String("sessionID", session.UUID.String()))
+	}
+	return nil
+}
+
+// Log in a user
+func (sessionService *sessionServiceImpl) LogUserIn(user *models.User, response http.ResponseWriter) httperrors.HTTPError {
+	dur := sessionService.sessionConfiguration.GetSessionDuration()
+	fmt.Println(dur)
```
Comment (Zorin95670) [inline services/sessionservice/session.go:217]: dur ? println ? :eyes:

File: services/sessionservice/session.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+	sess := newSession(user.ID, dur)
```
Comment (Zorin95670) [inline services/sessionservice/session.go:218]: sess :eyes:

File: services/sessionservice/session.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	}
+	CreateAndSetAccessTokenCookie(response, sess.UUID.String())
+	return nil
+}
+
+// Log out a user.
+func (sessionService *sessionServiceImpl) LogUserOut(sessionClaims *SessionClaims, response http.ResponseWriter) httperrors.HTTPError {
+	session := sessionService.get(sessionClaims.SessionUUID)
+	if session == nil {
+		panic("The logout function SHOULD only be called when a user is authenticated")
```
Comment (Zorin95670) [inline services/sessionservice/session.go:231]: not sure we want to panic in this case

File: http_support_test.go [hunk 12]
```diff
@@ -49,3 +50,31 @@ func (t *TestContext) assertResponseFieldIsEquals(field string, expectedValue st
 func assertValue(value string, expectedValue string) bool {
 	return expectedValue == value
 }
+
+func (t *TestContext) iSigninAsWithPassword(username, password string) error {
```
Comment (Zorin95670) [inline http_support_test.go:54]: Remove this as we discuss

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: services/sessionservice/session.go (hunk 9 above)
Comment (slashformotion) [reply services/sessionservice/session.go:217]: removed

File: persistence/repository/CRUDRepositoryImpl.go [hunk 13]
```diff
@@ -126,6 +142,11 @@ func (repository *CRUDRepositoryImpl[T, ID]) GetAll(sortOptions ...pagination.So
 	return entities, nil
 }
 
+func buildClauseFromSortOption(sortOption SortOption) clause.OrderByColumn {
+	clause := clause.OrderByColumn{Column: clause.Column{Name: sortOption.Column()}, Desc: sortOption.Desc()}
```
Comment (Zorin95670) [inline persistence/repository/CRUDRepositoryImpl.go:146]: `clause := ` to `return clause.(...)`
Missing documentation

File: services/sessionservice/session.go [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+		sessionService.logger.Warn("Rolled session",
+			zap.Uint("userID", session.UserID),
+			zap.String("sessionID", session.UUID.String()))
+	}
+	return nil
+}
+
+// Log in a user
+func (sessionService *sessionServiceImpl) LogUserIn(user *models.User, response http.ResponseWriter) httperrors.HTTPError {
+	dur := sessionService.sessionConfiguration.GetSessionDuration()
```
Comment (Zorin95670) [inline services/sessionservice/session.go:221]: what is dur ?

File: services/sessionservice/session.go [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+	defer sessionService.mutex.Unlock()
+	var i int
+	for sessionUUID, session := range sessionService.cache {
+		if session.IsExpired() {
+			// Delete the session in the database
+			err := sessionService.sessionRepository.Delete(session)
+			if err != nil {
+				panic(err)
+			}
+			// if the deletion of the session in the database was successfull,
```
Comment (Zorin95670) [inline services/sessionservice/session.go:159]: successful

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: features/basic_auth.feature [hunk 16]
```diff
@@ -0,0 +1,35 @@
+Feature: Login as superadmin using the basic authentification
+
+Scenario: Server should let us login/ with correct credentials
+  Given I request "/login" with method "POST" with json
+  | key      | value | type    |
+  | email  | admin-no-reply@badaas.com | string  |
+  | password | admin  | string |
+  Then I expect status code is "200"
+
```
Comment (Zorin95670) [inline features/basic_auth.feature:9]: remove extra line

File: features/basic_auth.feature [hunk 17]
```diff
@@ -0,0 +1,35 @@
+Feature: Login as superadmin using the basic authentification
```
Comment (Zorin95670) [inline features/basic_auth.feature:1]: this files is not lint and correctly indented

File: features/basic_auth.feature [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+Scenario: Server should let us login/ with correct credentials
```
Comment (Zorin95670) [inline features/basic_auth.feature:3]: Should be a success on valid credentials

File: features/basic_auth.feature [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+  Given I request "/login" with method "POST" with json
+  | key      | value | type    |
+  | email  | admin-no-reply@badaas.com | string  |
+  | password | admin  | string |
+  Then I expect status code is "200"
+
+
+Scenario: Server return an error on invalid credentials
```
Comment (Zorin95670) [inline features/basic_auth.feature:11]: Should be an error on invalid credentials

File: features/basic_auth.feature [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  Given I request "/login" with method "POST" with json
+  | key      | value | type    |
+  | email  | admin-no-reply@badaas.com | string  |
+  | password | wrongpassword  | string |
+  Then I expect status code is "401"
+  And I expect response field "err" is "wrong password"
+  And I expect response field "msg" is "the provided password is incorrect"
+  And I expect response field "status" is "Unauthorized"
+
+Scenario: Server should let us login and logout
```
Comment (Zorin95670) [inline features/basic_auth.feature:21]: Should be a sucess if we logout after a successful login

File: features/basic_auth.feature [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  Given I request "/login" with method "POST" with json
+  | key      | value | type    |
+  | email  | admin-no-reply@badaas.com | string  |
+  | password | admin  | string |
+  Then I expect status code is "200"
+
+
+Scenario: Server should return an error on logout without login first
```
Comment (Zorin95670) [inline features/basic_auth.feature:29]: Should be an error if we try to logout without login first

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Pull request merged**